src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java

```diff
@@ -103,6 +118,249 @@ protected void modifyCharacteristics(Branch<?> branch, BranchModificationInfos b
+                ((RatioTapChangerAdder) adder)
+                        .setLowTapPosition(tapChangerModificationInfos.getLowTapPosition().getValue());
+            } else {
+                ((PhaseTapChangerAdder) adder)
+                        .setLowTapPosition(tapChangerModificationInfos.getLowTapPosition().getValue());
+            }
+        } else if (tapChanger != null) {
+            if (adder instanceof RatioTapChangerAdder) {
+                ((RatioTapChangerAdder) adder)
+                        .setLowTapPosition(tapChanger.getLowTapPosition());
+            } else {
+                ((PhaseTapChangerAdder) adder)
+                        .setLowTapPosition(tapChanger.getLowTapPosition());
+            }
+        }
+        if (tapChangerModificationInfos.getTapPosition() != null && tapChangerModificationInfos.getTapPosition().getValue() != null) {
+            Integer oldTapPosition = tapChanger != null ? tapChanger.getTapPosition() : null;
+            tapChangerReports.add(ModificationUtils.getInstance().buildModificationReportWithIndentation(oldTapPosition,
+                tapChangerModificationInfos.getTapPosition().getValue(), "Tap position", 2));
+            if (adder instanceof RatioTapChangerAdder) {
+                ((RatioTapChangerAdder) adder)
+                        .setTapPosition(tapChangerModificationInfos.getTapPosition().getValue());
+            } else {
+                ((PhaseTapChangerAdder) adder)
+                        .setTapPosition(tapChangerModificationInfos.getTapPosition().getValue());
+            }
+        } else if (tapChanger != null) {
+            if (adder instanceof RatioTapChangerAdder) {
+                ((RatioTapChangerAdder) adder)
+                        .setTapPosition(tapChanger.getTapPosition());
+            } else {
+                ((PhaseTapChangerAdder) adder)
+                        .setTapPosition(tapChanger.getTapPosition());
+            }
+        }
+    }
+
+    private <T> void addTapChanger(TapChangerModificationInfos tapChangerModificationInfos, TapChanger<?, ?> tapChanger, T adder, Reporter subReporter) {
+        List<Report> tapChangerReports = new ArrayList<>();
+        modifyTapPositions(tapChangerModificationInfos, adder, tapChanger, tapChangerReports);
+
+        // Add steps
+        if (tapChangerModificationInfos.getSteps() != null) {
+            tapChangerReports.add(Report.builder()
+                .withKey("tapsModification")
+                .withDefaultMessage("        Taps")
+                .withSeverity(TypedValue.INFO_SEVERITY)
+                .build());
+            addTapchangerSteps(tapChangerReports, tapChangerModificationInfos, adder);
+        } else if (tapChanger != null) {
+            for (Object step : tapChanger.getAllSteps().values()) {
+                if (step instanceof RatioTapChangerStep) {
+                    RatioTapChangerStep ratioStep = (RatioTapChangerStep) step;
```

Comment:
this line can be deleted if you use:  
if (step instanceof RatioTapChangerStep ratioStep) {